src/microbe_stage/MembraneWaterRipple.cs

```diff
@@ -0,0 +1,666 @@
+                    if (movement.LengthSquared() > 0.001f)
+                    {
+                        lastValidDirection = currentDirection;
+                    }
+                }
+            }
+        }
+        else
+        {
+            currentSpeed = Mathf.Lerp(currentSpeed, 0.05f, 0.01f);
+
+            // If direction hasn't changed for a while but we're moving, nudge it
+            if (directionChangeTimer > 1.0f && wasMovingLastFrame)
+            {
+                // Force direction update if it's been frozen
+                Vector3 forwardVector = positionTarget.GlobalTransform.Basis.Z;
+                Vector2 orientationDir = new Vector2(forwardVector.X, forwardVector.Z).Normalized();
+                currentDirection = currentDirection.Lerp(orientationDir, 0.05f).Normalized();
+                directionChangeTimer = 0.0f;
+            }
+        }
+
+        // Make sure direction is always normalized
+        if (currentDirection.LengthSquared() < 0.01f)
+        {
+            currentDirection = lastValidDirection;
+        }
+
+        waterMaterial.SetShaderParameter(movementDirectionParam, currentDirection);
+        waterMaterial.SetShaderParameter(movementSpeedParam, currentSpeed);
+    }
+
+    /// <summary>
+    ///   Updates the local position array for the shader
+    ///   based on the real position history
+    /// </summary>
+    private void UpdateLocalPositions()
+    {
+        if (positionTarget == null || !positionTarget.IsInsideTree())
+            return;
+
+        Vector3 currentPos = positionTarget.GlobalPosition;
+
+        // Calculate local positions relative to current position
+        int count = isPositionHistoryFull ? PositionHistorySize : currentPositionIndex;
+
+        // Optimize based on camera distance so uses fewer samples for distant objects
+        int step = 1;
+        if (lastCameraDistance > 120.0f)
+            step = 3;
+        else if (lastCameraDistance > 80.0f)
+            step = 2;
+
+        // Access circular buffer with wrapping index
+        // Transforms world positions to local XZ offsets for shader
+        for (int i = 0; i < count; i += step)
+        {
+            int index = (currentPositionIndex - 1 - i).PositiveModulo(PositionHistorySize);
+
+            Vector3 offset = positionHistory[index] - currentPos;
+            Vector2 position = new Vector2(offset.X, offset.Z);
+
+            // Update both arrays ensuring array bounds
+            int targetIndex = i / step;
+            if (targetIndex < pastPositions.Length && targetIndex < godotPastPositions.Count)
+            {
+                pastPositions[targetIndex] = position;
+                godotPastPositions[targetIndex] = position;
+            }
+        }
+
+        // Fill remaining positions with zeros for lower detail sampling
+        if (step > 1)
+        {
+            for (int i = count / step + 1; i < count && i < pastPositions.Length && i < godotPastPositions.Count; ++i)
```

Comment:
Doesn't this end early? If any of the lengths are not the same, then this will fill only positions with zero until the first length check that fails?

I really think you ought to make a constant for how many elements each of the positions arrays must always have and then rewrite the code a bit to take advantage of that to simplify the bounds checking and ensure there won't be strange bugs if someone ever changes the size of one of the arrays.